blog/en-us/spring-boot-dubbo-start-stop-analysis.md

```diff
@@ -0,0 +1,223 @@
+
+```java
+@SpringBootApplication(scanBasePackages = "com.alibaba.boot.dubbo.demo.consumer.controller")
+public class DubboConsumerDemo {
+
+    public static void main(String[] args) {
+        SpringApplication.run(DubboConsumerDemo.class,args);
+    }
+
+}
+```
+
+In fact, if you want to answer such a question. We need to abstract this problem firstly, that is, a JVM process, under what circumstances will it exit?
+
+Take Java 8 as an example. By referring to the JVM language specification[1], there is a clear description in Section 12.8:
+
+> A program terminates all its activity and *exits* when one of two things happens:
+>
+> - All the threads that are not daemon threads terminate.
+> - Some thread invokes the `exit` method of class `Runtime` or class `System`, and the `exit` operation is not forbidden by the security manager.
+
+Therefore, in view of the above situation, we judge that there must be some non-daemon thread not exiting. All thread information can be seen by `jstack`, including whether they are daemon threads, and `jstack` can be used to find out which threads are non-deamon.
+
+```sh
+➜  jstack 57785 | grep tid | grep -v "daemon"
+"container-0" #37 prio=5 os_prio=31 tid=0x00007fbe312f5800 nid=0x7103 waiting on condition  [0x0000700010144000]
+"container-1" #49 prio=5 os_prio=31 tid=0x00007fbe3117f800 nid=0x7b03 waiting on condition  [0x0000700010859000]
+"DestroyJavaVM" #83 prio=5 os_prio=31 tid=0x00007fbe30011000 nid=0x2703 waiting on condition  [0x0000000000000000]
+"VM Thread" os_prio=31 tid=0x00007fbe3005e800 nid=0x3703 runnable
+"GC Thread#0" os_prio=31 tid=0x00007fbe30013800 nid=0x5403 runnable
+"GC Thread#1" os_prio=31 tid=0x00007fbe30021000 nid=0x5303 runnable
+"GC Thread#2" os_prio=31 tid=0x00007fbe30021800 nid=0x2d03 runnable
+"GC Thread#3" os_prio=31 tid=0x00007fbe30022000 nid=0x2f03 runnable
+"G1 Main Marker" os_prio=31 tid=0x00007fbe30040800 nid=0x5203 runnable
+"G1 Conc#0" os_prio=31 tid=0x00007fbe30041000 nid=0x4f03 runnable
+"G1 Refine#0" os_prio=31 tid=0x00007fbe31044800 nid=0x4e03 runnable
+"G1 Refine#1" os_prio=31 tid=0x00007fbe31045800 nid=0x4d03 runnable
+"G1 Refine#2" os_prio=31 tid=0x00007fbe31046000 nid=0x4c03 runnable
+"G1 Refine#3" os_prio=31 tid=0x00007fbe31047000 nid=0x4b03 runnable
+"G1 Young RemSet Sampling" os_prio=31 tid=0x00007fbe31047800 nid=0x3603 runnable
+"VM Periodic Task Thread" os_prio=31 tid=0x00007fbe31129000 nid=0x6703 waiting on condition
+
+```
+
+> We can find all the thread digests by `grep tid` here, and find the line that doesn't contain the daemon keyword by `grep -v` command.
+
+We can get some information by the above results:
+
+- There are two threads `container-0`, `container-1` is very suspicious, Ther are non-daemon threads in wait state.
+- There are alse some threads about GC, and threads that start with `VM`, They are also non-daemon threads, but they are most likely the JVM's own threads, which are ignored here.
```

Comment:
- There are alse some threads about GC, and threads that start with `VM`. They are also some non-daemon threads, but they are most likely the JVM's own threads, which we can ignore for now.